### PR TITLE
Realtime spellchecking big tuneup

### PR DESCRIPTION
**Pull request by adamconroy**
This PR is a collection of most of the tune up requests that we've figured out in the past few weeks:
+ suppress gutter tooltips
+ ignore for this document relint
+ add to user dict relint
+ only lint what's on screen and on scroll
+ change underline color (it is now a medium brownish yellow that shows up ok on both dark and light backgrounds, medium grey is kind of tough to see but hopefully not many people are using that)
+ only relint the line that was just typed
+ add sane caps to what is processed and set it in ui prefs
+ run lint on opening up a project with already saved files tune up

What wasn't done and why: `suppress underline mouseover tooltip`
So, I actually _did_ do this by removing the annotation from the marker and tracking them separately as that's what triggers the tooltip. Unfortunately this causes some REALLY weird interactions with ace.js and how we manage the markers. It also adds some extra complexity to track that just feels messy. The ideal here is to patch ace.js to allow annotations that don't trigger tooltips. These tooltips can't be trivially suppressed like the gutter ones because they're actually browser tooltips and not client created elements. Also, the tooltips are super flaky and don't update properly so I've changed them to just say "spellcheck" instead of the word you're mousing over. It'll be a nice enhancement to do if we decide to later on to actually get these puppies suppressed because they're unreliable anyway.

### CODE REVIEW

**Comment by adamconroy** [comment]
Oh also, I fixed https://github.com/rstudio/rstudio/issues/4757 in here because I couldn't leave well enough alone. Somehow `memoized[word]` was set but returning undefined and throwing an exception in typo.js.

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditorWidget.java [hunk 1]
```diff
@@ -1098,48 +1118,102 @@ public void execute()
...
+
+   public void removeMarkersAtCursorPosition()
+   {
+      // Defer this so other event handling can update anchors etc.
+      Scheduler.get().scheduleDeferred(() ->
+      {
+         Position cursor = editor_.getCursorPosition();
+         JsArray<AceAnnotation> newAnnotations = JsArray.createArray().cast();
+
+         for (int i = 0; i < annotations_.size(); i++)
```
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditorWidget.java]: It looks like this code is more or less duplicated three times -- is it straightforward to share implementations?

Comment (adamconroy) [reply src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/AceEditorWidget.java]: It was! I was able change up the predicate code in a way that was reusable. I wanted to refactor it last night but was a little stumped until I realized I could just use one of the functions already there.

**Comment by adamconroy** [comment]
@kevinushey would love your review on this :)

File: src/gwt/src/org/rstudio/studio/client/workbench/prefs/model/UIPrefsAccessor.java [hunk 2]
```diff
@@ -494,6 +494,16 @@ public String getThemeErrorClass()
                     JsArrayString.createArray().<JsArrayString>cast());
    }
 
+   public PrefValue<Integer> maxCheckWords()
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/prefs/model/UIPrefsAccessor.java]: Does this need some code in `UIPrefs.java` to update the preference after the user has changed the option? See e.g. https://github.com/rstudio/rstudio/commit/72c67f5317134cff6911391d88a1604ec13237d8#diff-525e35862e06774c892a1991ec1784b0.

Is an associated change in `TextEditingTarget.java` also warranted (to ensure that existing documents see the updated preference value)?

Comment (adamconroy) [reply src/gwt/src/org/rstudio/studio/client/workbench/prefs/model/UIPrefsAccessor.java]: Oh maybe we do, thanks! I'll toss that in.

**Pull request merged**